server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java

```diff
@@ -0,0 +1,172 @@
+  private volatile boolean started = false;
+  private volatile boolean initialized = false;
+  private final ExecutorService executor = Execs.singleThreaded("CachingCostBalancerStrategy-executor");
+  private final ClusterCostCache.Builder clusterCostCacheBuilder = ClusterCostCache.builder();
+
+  @Inject
+  public CachingCostBalancerStrategyFactory(ServerInventoryView serverInventoryView)
+  {
+    this.serverInventoryView = Preconditions.checkNotNull(serverInventoryView);
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (lifecycleLock) {
+      if (executor.isShutdown()) {
+        throw new ISE("CachingCostBalancerStrategyFactory has been stopped");
+      }
+      if (started) {
+        throw new ISE("CachingCostBalancerStrategyFactory is already started");
+      }
+
+      serverInventoryView.registerSegmentCallback(
+          executor,
+          new ServerView.SegmentCallback()
+          {
+            @Override
+            public ServerView.CallbackAction segmentAdded(
+                DruidServerMetadata server, DataSegment segment
+            )
+            {
+              clusterCostCacheBuilder.addSegment(server.getName(), segment);
+              return ServerView.CallbackAction.CONTINUE;
+            }
+
+            @Override
+            public ServerView.CallbackAction segmentRemoved(
+                DruidServerMetadata server, DataSegment segment
+            )
+            {
+              clusterCostCacheBuilder.removeSegment(server.getName(), segment);
+              return ServerView.CallbackAction.CONTINUE;
+            }
+
+            @Override
+            public ServerView.CallbackAction segmentViewInitialized()
+            {
+              initialized = true;
+              return ServerView.CallbackAction.CONTINUE;
+            }
+          }
+      );
+
+      serverInventoryView.registerServerCallback(
+          executor,
+          server -> {
+            clusterCostCacheBuilder.removeServer(server.getName());
+            return ServerView.CallbackAction.CONTINUE;
+          }
+      );
+
+      started = true;
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lifecycleLock) {
+      if (executor.isShutdown()) {
+        throw new ISE("CachingCostBalancerStrategyFactory has been already stopped");
+      }
+      if (!started) {
+        throw new ISE("CachingCostBalancerStrategyFactory is not started");
+      }
+      executor.shutdownNow();
+    }
+  }
+
+  @Override
+  public BalancerStrategy createBalancerStrategy(final ListeningExecutorService exec)
+  {
+    if (!started) {
+      throw new ISE("CachingCostBalancerStrategyFactory need to be started");
+    }
+    if (executor.isShutdown()) {
+      throw new ISE("CachingCostBalancerStrategyFactory has been stopped");
+    }
+    if (initialized) {
+      try {
+        CompletableFuture<CachingCostBalancerStrategy> future = CompletableFuture.supplyAsync(
```

Comment:
It's updated only from Curator, likely from a single thread, so I don't think it will make any performance change. OTOH it allows to avoid nasty threading and error-catching logic in the factory.